LSP19 Social Registry (#154)
lukso-network/LIPs: LSPs/LSP-19-Social-Registry.md

      "key": "0xaa093cc8e40d9473239c9fefe0f7e73ad8b9fb1bfca176dcaf5a8af4eacfb1f4",
      "keyType": "Singleton",
      "valueType": "bytes",
      "valueContent": "JSONURL"
  }
```

This registry should be updated everytime a new post is added by the user.
Note: It might not be necessary to update the registry for each follow, like or dislike. A balanced trade-off between interoperability/decentralization and cost/UX could be found.

The linked JSON file SHOULD have the following format:

```js
{
  "LSP19SocialRegistry": {
    "posts": [ // Messages authored by the profile. Includes original posts, comments and reposts.
      {
        "url": "String", // The url in decentralized storage with the post content and metadata
        "hash": "Bytes32" // The hash of the post object
      },
      ...
    ],
    "follows": [ "Address", ... ], // UPs this account has subscribed.  Will compose the account's feed.
    "likes": [
      {
        "url": "String", // The url in decentralized storage with the post content and metadata
        "hash": "Bytes32" // The hash of the post object
      }
    ], // The identifier (hash) of all the posts this account has liked,

>     "dislikes": [

Reviewer: Facebook consciously decided against dislikes. It would be good to look into that reasoning before adding this here.

Author: We also did not use it in our Lookso project, as we don't find it relevant. However, we thought it might be important proposing a really complete and flexible standard, in order for people to be able to build any kind of social app they want on top of that. For instance, it might be relevant for people doing a Youtube or Reddit like DApp

---
lip: 19
title: Social Registry
status: Draft
type: LSP
author: Samuel Videau <samuel@dropps.io>, António Pedro <antonio@dropps.io>
created: 2022-07-26
updated: 2023-01-01
requires: ERC725Y, LSP2
---

## Simple Summary

This standard describes a smart contract, and a data model to store Social Media information such as posts, likes and follows.

## Abstract

> This standard defines a set of data formats and a key-value pair to create a Social Media Feed, combining [ERC725Account](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-0-ERC725Account.md) and an open distributed storage network such as [IPFS](https://ipfs.tech/) or [ARWEAVE](https://arweave.org).

Reviewer 2: Maybe?

```suggestion
This standard defines a set of data formats and a key-value pair to create a decentralised Social Media Feed, combining [LSP0-ERC725Account](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-0-ERC725Account.md) and an open distributed storage network such as [IPFS](https://ipfs.tech/) or [ARWEAVE](https://arweave.org).
```

> It also defines a smart contract used to guaranty authenticity and timestamp of a post.

Reviewer 2: ```suggestion
It also defines a smart contract that can be used to guaranty authenticity and timestamp of a post.
```

> ## Motivation

Reviewer 2: Super the motivation section 👌  Very on point!

Real interoperability requires social media itself to be separated from social media companies. This proposal aims to create a common interoperable standard in which messages generated on one social media app could be transported and read in any other application.

Using a standardized data model to store social media makes content platform-independent and allows it to be read and stored easily. This content can be added to an [ERC725Account](https://github.com/lukso-network/LIPs/blob/main/LSPs/LSP-0-ERC725Account.md), giving it a Social Media Account character.

## Specification

### LSP19SocialRegistry

> A Universal Profile's Social Media State will live under a record referenced by the "LSP19SocialRegistry" data key of their ERC725Y store.

Reviewer 2: Maybe we can keep it generic here? As Universal Profile could be one of the type of contracts that use it, but we don't know yet which other type of contracts could use it.

```suggestion
A Social Media State (for instance the one of a Universal Profile) will live under a record referenced by the "LSP19SocialRegistry" data key of their ERC725Y store.
```

The JSON Url stored inside points to a JSON file that lists all the social media actions of a profile, including posts, likes, dislikes and follows.

```json
  {
      "name": "LSP19SocialRegistry",
      "key": "0xaa093cc8e40d9473239c9fefe0f7e73ad8b9fb1bfca176dcaf5a8af4eacfb1f4",
      "keyType": "Singleton",
      "valueType": "bytes",
      "valueContent": "JSONURL"
  }
```

> This registry should be updated everytime a new post is added by the user.

Reviewer 2: Let's ensure the standard verbs are in uppercase.

```suggestion
This registry SHOULD be updated everytime a new post is added by the user.
```

> Note: It might not be necessary to update the registry for each follow, like or dislike. A balanced trade-off between interoperability/decentralization and cost/UX could be found.

Reviewer 2: ```suggestion
> Note: It might not be necessary to update the registry for each follow, like or dislike. A balanced trade-off between interoperability/decentralization and cost/UX could be found.
```

      }
    ], // The identifier (hash) of all the posts this account has liked,
    "dislikes": [
      {
        "url": "String", // The url in decentralized storage with the post content and metadata
        "hash": "Bytes32" // The hash of the post object
      }
    ], // The identifier (hash) of all the posts this account has disliked
  }
}
```

Below is an example of a social registry:

```JSON
{
  "posts": [
    {
      "url": "ar://NUb9WJ9BbbxLfIDgGwq4zPECBz_df0CrhBmRsYsn8-Y",
      "hash": "0xb1029df66ea5ae5cdcc0e84b6e048e37b3df14a4aec92fc4c23d86f8c62e4a4c"
    }
  ], 
  "likes": [
    {
      "url": "ar://oHJwvoggzfUxv2WpJIeIlLNC-OR1X1CmGO2zp7BeRgk",
      "hash": "0x50f7488034e24cf441d5d02a174d5f56930dbbdbb8815dbb54346be1c5648377"
    }
  ],
  "dislikes": [
    {
      "url": "ar://PPJDYJf6AgZZYKOXZtBBgmdz_-XFCFaBiTuh7Mojecc",
      "hash": "0xe15aaaa78c05fc9f9f6d9099db9e984dffe7a732e4fdd503187ad1a91f8390c7"
    }
  ],
  "follows": [
    "0x53529E4164E5CCA7d6A1C55f8500A57D0F435bee"
  ]
}
```

### Profile Posts

A Profile Post can be an original message, a comment on another post or a repost. The JSON file should have the following format:

> Not all fields are required. For example, a `repost` doesn't need a message, but it should have the `url` and `hash` to the original post.

Reviewer 2: I cannot find the `repost` field below. Where is it located?

Also for the `repost`, is it an absolute requirement that it has the `url` and `hash`? I guess yes then in this case, the wording would be MUST here.

```suggestion
**Not all fields are required**. For example, a `repost` doesn't need a message, but it MUST have the `url` and `hash` to the original post.
```

Author: Maybe that would need additional explanation, but basically a repost is a post that have a `childPost` value filled

> The `LSP19ProfilePostSignature` property is optional and depends on the use case. It is used to authenticate a post through a controller EOA (Externally Owned Account)

Reviewer 2: Does it has to be an EOA? Can't it be another 🆙 ? 😉 

```suggestion
The `LSP19ProfilePostSignature` property is OPTIONAL and depends on the use case. It is used to authenticate a post through a controller EOA (Externally Owned Account)
```

Author: Hmm what do you mean ?

```js
{
  "LSP19ProfilePost": {

>     "version": "0.0.1", // The Metadata version of this post

Reviewer 2: Might be good to add a sentence in the standard to describe this field.

    "author": "Address", // The Universal Profile who authored the post

>     "locale": "string", // language code - Country Code (de_DE)

Reviewer 2: Might be good to be consistent. Above, strings are mentioned as `String` (uppercase `S`).

<img width="434" alt="image" src="https://github.com/lukso-network/LIPs/assets/31145285/b0923ab3-0a00-4fdd-b0ff-b923ece00a91">

    "app": "string", // The platform that originated this post
    "validator": "Address", // Address of a validator smart contract which will authenticate a post and provide its publication date (more info bellow)
    "nonce": "string", // Random value to allow duplicates
    "message": "string", // The post original content
    "links": [
      {
        "title": "string", // The link's label
        "url": "string"
      },
      ...
    ],
    "tags": [ // Tags attached to a post
      "string",
      ...
    ],
    "medias": [ // Medias attached to a post
      {
        "hashFunction": "keccak256(bytes)",
        "hash": "string",
        "url": "string", 
        "fileType": "string"
      }
    ],
    "assets": [

>       "interface": "string" // Contract interface

Reviewer 2: Maybe here you could provide the option to be either:

- a contract interface name (e.g: `ILSPN...`)
- or a bytes4 interface identifier (e.g: `0x3e89ad98` for LSP0 interface).

```suggestion
      "interface": "string | bytes4" // Contract interface
```

Author: Actually it might maybe be good having both

The `postWithJsonUrl` function extends the `post` function by allowing the UP to also specify a reference to the latest Social Media Record in the `jsonUrl` argument. 
This function first verifies that the UP implements the ERC725Y standard (which includes a key/value store) and then sends a transaction to the UP to update the registry reference in the UP's key/value store with the `jsonUrl` value.

### Implementation

```solidity
// SPDX-License-Identifier: MIT

pragma solidity ^0.8.7;

import { Context } from "@openzeppelin/contracts/utils/Context.sol";
import {_INTERFACEID_ERC725Y} from "@erc725/smart-contracts/contracts/constants.sol";
import { OwnableUnset } from "@erc725/smart-contracts/contracts/custom/OwnableUnset.sol";
import { ERC165Checker } from "@openzeppelin/contracts/utils/introspection/ERC165Checker.sol";
import { ILSP6KeyManager} from "@lukso/lsp-smart-contracts/contracts/LSP6KeyManager/ILSP6KeyManager.sol";

/**
* @title LSP19 post validator
* @notice A validator tailored for Universal Profiles and content publishing
* @dev Writes to the Universal Profile key/value store
*/

> contract LSP19PostValidator is Context {

Reviewer 2: Is there a standard interface for the contract (some standard functions), or it could be implemented in any way.

I am asking as the title says:

<img width="347" alt="image" src="https://github.com/lukso-network/LIPs/assets/31145285/019db9da-366a-4dd8-b0bf-6a394731475b">

And from my understanding, I thought it would also include an interface (set of functions).
